fi: fix behaviour of S_2_TO_HEAVY (adding secondary stress) and code readability changes

jaacoppi opened the pull request:
Stress flag S_2_TO_HEAVY is currently only used by finnish.

Current behaviour skips adding secondary stress if the following syllable is heavy. The behaviour should be to skip adding secondary stress if the rest of the word (excluding last syllable) contains a heavy syllable.

behaviour of espeak-ng -x before commit:
p'apIlj,ot:I
k'ilIs,ev&
k'eskust,elIsIn
m'atemat,i:kass2,anI
t'apa:mIs,esta
k'eskust,elIs,it:e
'opIsk,elIj,oina
p'erust,elem,at:om,alta

after commit:
p'apIlj,ot:I
k'ilIs,ev&
k'eskust,elIsIn
m'atemat,i:kass2,anI
t'apa:mIs,esta
k'eskustelIs,it:e
'opIskelIj,oina
p'erustelem,at:om,alta

The commit changes the behaviour of last three words.

Source of grammar rule and examples of expected behaviour: http://scripta.kotus.fi/visk/sisallys.php?p=13

rhdunn commented:
Generally, this looks good. Some comments:
1.  There should be braces around the for loop in the first patch as it spans more than one line, even though it only consists of a single statement. This is to aid readability.
2.  I would add `STRESS_` to the start of the define names, to make it clearer what they are referring to and to help avoid potential name collisions.
3.  I generally prefer separate changes to be split out into separate patches (e.g. adding the defines, and using bool in the loop control). This makes it easier to see what each change does, and helps investigate problems using tools like `git bisect`.

__NOTE:__ There are some stress constants in synthesize.h. It would be useful to use those consistently. Then rename then to `STRESS_IS_DIMINISHED`, etc. This is to provide consistency between compiling the data files, reading the data files, and processing the data to generate the speech.

jaacoppi commented:
Okay, I' have:
1. Reverted code readability changes
2. Added braces around the for loop
3. Included a new commit for stress rules of finnish numerals

I'll continue  with the code readability changes later. I can make a new pull request after this one  if that's better. Does it matter if if use STRESS_DIMINISHED or STRESS_IS_DIMINISHED? @rhdunn you used both examples in your previous comment.

jaacoppi commented:
Here's the stress numbers used in dictionary.c:
https://github.com/espeak-ng/espeak-ng/blob/3dea33965eb023d1589e6a716cd1ac54a5c01e2d/src/libespeak-ng/dictionary.c#L1050-L1055

And here's synthdata.h: https://github.com/espeak-ng/espeak-ng/blob/3dea33965eb023d1589e6a716cd1ac54a5c01e2d/src/libespeak-ng/synthesize.h#L321-L326

Are they supposed to mean the same thing? Which variable names are most suitable?  dictionary.c doesn't have a 2, but  has a 5 for priority stress in the code..

rhdunn commented:
I think `STRESS_IS_...` is the better prefix, given the `is...` naming in `synthesize.h`.

They should mean the same thing. Also, `phsource/phonemes` has:

    %%    : stress_type 0 - diminished stress
    %     : stress_type 1 - unstressed
    ,     : stress_type 2 - secondary stress
    =     : stress_type 2 - place stress on previous vowel
    ,,    : stress_type 3 - secondary stress
    '     : stress_type 4 - primary stress
    ''    : stress_type 5 - primary stress (priority)
    '!    : stress_type 6 - primary stress (emphasized)

I would make the names consistent across the different places they are used/referenced. Something like:

    // stress type
    #define STRESS_IS_DIMINISHED 0 // 'diminished'
    #define STRESS_IS_UNSTRESSED 1 // 'unstressed' -- unstressed, or weak stress
    #define STRESS_IS_DEFAULT    2 // 'default' -- not stressed
    #define STRESS_IS_SECONDARY  3 // 'secondary'
    #define STRESS_IS_PRIMARY    4 // 'primary' -- primary, or main stress
    #define STRESS_IS_PRIORITY   5 // 'priority'
    #define STRESS_IS_EMPHASIZED 6 // 'emphasized'

__NOTE:__ You can use `STRESS_IS_NOT_STRESSED` and `not_stressed` or `not-stressed` for stress type `2` instead of `default` if you think that is more understandable.

__NOTE:__ The names in `'` are the suggested names for the `stress_type` option, if you decide to implement that as well. You can use `MNEM_TAB` and `LookupMnem` to implement that functionality, like in `voices.c`.

rhdunn commented:
Thanks for the fixes. Some additional comments:
1. Your name is set to `=` in the commit logs. Is this intentional? It was "Juho Hiltunen" previously.
1. You don't need to revert the previous commits -- either reset to the previous commit and redo them, or build the fixes on top of the existing commits.
1. You don't need to reference commits that are being fixed, or review comment discussions in the commit messages.

With those changes, I will merge the commits.

Another thing -- it is easier if you keep different things (e.g. the code changes, and the fi language changes) as separate branches. This makes it easier for me to review and merge the different changes, and makes it easier to redo changes like the ones discussed here.

jaacoppi commented:
Okay, thanks.
I'll close this pull request and add two new ones.
